# BinaRena

[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)
![javascript](https://badges.aleen42.com/src/javascript.svg)
![jasmine](https://badges.aleen42.com/src/jasmine.svg)

**Binarena** ("bin arena") is an interative visualizer of metagenomic assembly that facilitates observation, semi-manual and semi-automatic binning of metagenomic contigs.

## [live demo](https://qiyunlab.github.io/binarena/demo.html)

Please check out this [live demo](https://qiyunlab.github.io/binarena/demo.html). It is a fully functional program with a pre-loaded sample dataset. You may drag and drop your own data into the interface and start to analyze!


## (No)Installation

[Download](https://github.com/qiyunlab/binarena/archive/refs/heads/master.zip) this program, unzip, and double-click "**BinaRena.html**". That's it!

BinaRena is a client-end web application. It is written in vanilla JavaScript, with minimum usage of third-party libraries, which are all bundled in the package. it does not require installation, nor does it require a web server running in the backend. In other words, it is literally a webpage running in your browser.


## Mouse and keyboard shortcuts

The BinaRena interface is like a [digital map](https://www.google.com/maps) and is very intuitive. Although operations can be triggered through the graphical interface, the following basic operations are most convenient by using your mouse and keyboard.

- Drag the plot or use arrow keys (<code>&larr;</code>, <code>&uarr;</code>, <code>&rarr;</code> and <code>&darr;</code>) to move around.
- Zoom in/out: Mouse wheel, or use `=` and `-`.
- Click a contig to select it. Hold `Shift` and click to select multiple contigs.
- Click a selected contig to unselect it. Hold `Shift` and click to unselect multiple contigs.
- Press `Enter` to enter **polygon selection** mode. Then use mouse clicks to draw a polygon to contain multiple contigs. Press `Enter` again to complete selection.
- Press `Delete` or `Backspace` to **mask** selected contigs.
- Press `M` to toggle between selection and masking modes. In the latter scenario, polygon-selected contigs will be directly masked.
- Press `P` to take a screenshot of the current view.
- Press `0` to reset the plot view.
- Press `Space` to create a new bin from selected contigs.
- Press `.` (`>`) to add selected contigs to the current bin, press `,` (`<`) to remove selected contigs from the current bin, press `/` to update the current bin with selected contigs (i.e., replace its content).


## Input files

### Data table

The most typical input file format is a **data table** (tab-delimited file, or [TSV](https://en.wikipedia.org/wiki/Tab-separated_values)) that stores properties of contigs (scaffolds) in an assembly. Here is an [example](examples/input.tsv).

Each row represents one contig. The first column must be unique contig identifiers. The remaining columns are arbitrary properties (such as length, coverage, GC%, taxonomy, etc.). These fields may be in any of the following four types:

- number(`n`), category(`c`), feature(`f`), description(`d`)

The program will guess column type based on the data. To be safer, you may append a type code to a column name following a pipe (`|`), such as "`genes|n`" and "`plasmid|c`".

In a **feature**-type column, a cell can have multiple comma-separated features, such as "`Firmicutes,Proteobacteria,Cyanobacteria`".

One may assign a **weight** to each category or feature, using a numeric suffix following a colon (`:`), such as "`Ruminococcus:80`", and "`rpoB:1,dnaK:2,16S:5`".
This weight can represent quantity, proportion, confidence, etc.

**Missing** values are allowed. The program automatically identifies them, such as empty cell, "`-`", "`N/A`", "`na`", "`#NaN`", etc.

### Assembly file

An assembly file is a [FASTA](https://en.wikipedia.org/wiki/FASTA_format) format file containing multiple contig sequences. BinaRena can read DNA sequences and calculate **length** and **GC content** of each contig. If this file is generated by [SPAdes](https://cab.spbu.ru/software/spades/) or [MEGAHIT](https://github.com/voutcn/megahit), BinaRena can recognize **coverage** from sequence titles.


## Calculations

BinaRena implements several algorithms to calculate binning-related metrics. They can be launched from the "**Advanced calculation**" button in the widget bar.

### Silhouette score

The [silhouette coefficient](https://en.wikipedia.org/wiki/Silhouette_(clustering)) measures how similar a contig is to other contigs in the same bin, as in contrast to contigs in other bins. It ranges from -1 (worse) to 1 (best). The silhouette score, i.e., the mean silhouette coefficient of all contigs, evaluates the quality of a binning plan.

### Adjusted Rand index

The [adjusted Rand index](https://en.wikipedia.org/wiki/Rand_index) (ARI) measures the consistency between two binning plans. Higher is better. Two identical plans have ARI = 1. Two random plans will have an ARI close to 0.


## Contact

Please forward any questions to the project leader: **Dr. Qiyun Zhu** (qiyunzhu@gmail.com).
